xarray/core/computation.py

```diff
@@ -1843,36 +1846,96 @@ def where(cond, x, y, keep_attrs=None):
     )
 
 
-def polyval(coord, coeffs, degree_dim="degree"):
+@overload
+def polyval(coord: DataArray, coeffs: DataArray, degree_dim: Hashable) -> DataArray:
+    ...
+
+
+@overload
+def polyval(coord: T_Xarray, coeffs: Dataset, degree_dim: Hashable) -> Dataset:
+    ...
+
+
+@overload
+def polyval(coord: Dataset, coeffs: T_Xarray, degree_dim: Hashable) -> Dataset:
+    ...
+
+
+def polyval(
+    coord: T_Xarray, coeffs: T_Xarray, degree_dim: Hashable = "degree"
+) -> T_Xarray:
     """Evaluate a polynomial at specific values
 
     Parameters
     ----------
-    coord : DataArray
-        The 1D coordinate along which to evaluate the polynomial.
-    coeffs : DataArray
-        Coefficients of the polynomials.
-    degree_dim : str, default: "degree"
+    coord : DataArray or Dataset
+        Values at which to evaluate the polynomial.
+    coeffs : DataArray or Dataset
+        Coefficients of the polynomial.
+    degree_dim : Hashable, default: "degree"
         Name of the polynomial degree dimension in `coeffs`.
 
+    Returns
+    -------
+    DataArray or Dataset
+        Evaluated polynomial.
+
     See Also
     --------
     xarray.DataArray.polyfit
-    numpy.polyval
+    numpy.polynomial.polynomial.polyval
     """
-    from .dataarray import DataArray
-    from .missing import get_clean_interp_index
-
-    x = get_clean_interp_index(coord, coord.name, strict=False)
 
+    coeffs = coeffs.sortby(degree_dim)
```

Comment:
Ok, thats clever.

For further simplification the best would be to reversly order the coeffs (highest first, since thats how the loop goes).
But I run into the problem that `reindex` causes problems when no explicit index is existent in coeffs.

I don't actually know if this is intentional or a bug?
```python
d = xr.DataArray([1, 2, 3], dims="x")
d.reindex({"x": [2, 1, 0]})  # does not reorder the values
d.reindex({"x": [2, 0]})  # even raises a ValueError
```

Reply:
I think we can require that `degree_dim` have an explicit index, so that the statement is explicit. Since most users will be using the output of `polyfit` it'll be OK.